smtpd/smtp.c

```diff
@@ -297,14 +301,52 @@ smtp_accept(int fd, short event, void *p)
 		fatal("smtp_accept");
 	}
 
-	if (smtp_session(listener, sock, &ss, NULL, NULL) == -1) {
+	/* XXX: make sure this does not break anything, as this used to be after
```

Comment:
We need to discuss about this, I don't think it is the correct way to deal with this.

Basically you are implementing a protocol, smtp_session implements another one; and the smptf_session I'm working on implement another one.

I think the smtp_accept() function should only call *one* of them, in your case it should only call proxy_session() which would internally be in charge of calling smtp_session().

Moreover, I think we should have an smtp_engine.c file containing all primitives + a state machine so that you don't need to poke into smtp_session() ever.

Can we discuss this sometime over IRC ?